### ISSUE TITLE
ca.get doesn't work as advertised

### ISSUE THREAD

**Issue by brunoseivam** [pyepics/pyepics#141]
`ca.get` doesn't seem to work as advertised. The documentation says that a disconnected channel will return None as a value, but it actually throws.

```
python
Python 2.7.15rc1 (default, Nov 12 2018, 14:31:15) 
[GCC 7.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from epics import ca
>>> chid = ca.create_channel('dummy')
>>> ca.get(chid, timeout=1.0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/epics/ca.py", line 437, in wrapper
    name(chid), timeout))
epics.ca.ChannelAccessException: get() timed out waiting 'dummy' to connect (1 seconds)
>>> ca.get(chid, timeout=1.0, wait=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/epics/ca.py", line 437, in wrapper
    name(chid), timeout))
epics.ca.ChannelAccessException: get() timed out waiting 'dummy' to connect (1 seconds)
```

**Comment by brunoseivam** [comment]
@domitto

**Comment by newville** [comment]
@brunoseivam  Hm, yes I see that too.   At first, one might be inclined to view this as a documentation bug, but I sort of think the "return None" is preferable.   Do you or anyone else have a strong preference?  I doubt we'd be able to please everyone here. 

Maybe the right question is:  Does anyone object to treating this as a behavior bug (and not as documentation bug) and fixing `ca.get()` to return `None` for an unconnected channel?   

FWIW, I think the solution is just a matter of replacing the `withConnectedCHID` decorator with a `withCHID` decorator.

**Comment by danielballan** [comment]
Since you ask, I'll share our experience with this question in caproto. We experimented with always raising `TimeoutError` on timeout, never returning `None`. Anecdotally, in our use of caproto on the experimental floor at BNL so far, we have found raising to be beneficial. It tends to produce clearer error messages in common usages and does not remove any flexibility.

Downstream code often neglects to check that the return value `is not None` in every case, and therefore a `None` value tends to propagate through one or more calls, perhaps into other libraries, before eventually producing a confusing error. To cite one example we saw, the return value might be cached for a time and then later `put` back to the IOC. When the cached value is `None,` an error comes out of `put`, and the origin of the problem is obfuscated. Of course, code that is _able_ to tolerate a timeout and somehow work around it can do:

```
try:
    # get...
except TimeoutError:
    # try something else?
```

More often, for the use cases we have seen, the best thing to do in this situation is fail immediately to and alert the user that something is disconnected, so it is reasonable to make the code that really wants to _ignore_ a timeout actively mark that, as above.

I defer to @newville in every way regarding what is best for pyepics and the large volume of downstream user code that it supports. This is just my two cents, based on what we think we've learned from caproto so far. :- )

**Comment by brunoseivam** [comment]
This is our use case: we have a simple script that monitors the health of a number of IOCs. We know all the PVs that we want to `caget` beforehand. In this particular case, a disconnected PV is not exceptional: it is in fact something that we want to detect. We were using `caget_many` to accomplish this, but we were surprised by it throwing (because of `caget`'s current behavior). Currently, when `caget_many` throws, we can  only know that _one_ PV is disconnected.

Now, regarding the fix, while I'd love to have the documentation be right and the code wrong here, I think it would be dangerous to change the behavior of `caget`, especially not knowing how many people depend on it throwing. I tend to prefer more conservative changes with regards to public APIs, so I'd suggest that a flag be added to `ca.get`: `throw`, with default value `True`. This would be the same approach taken by [cothread](http://controls.diamond.ac.uk/downloads/python/cothread/2-3/docs/html/catools.html#cothread.catools.caget). That way existing code remains compatible.

**Comment by newville** [comment]
@danielballan @brunoseivam  Thanks!  A few more thoughts (but I have no strong opinion on this).  

I do generally agree that "raise early" is usually the better approach and that "return None" is often just deferring the problem ("punting").   OTOH, None cannot be a valid PV value, so it is a reliable signal (of something, I guess) too.

It's not obvious to me whether an unconnected channel should be considered an exception.  Networks are unreliable, and the lack of connection may be a temporary condition and may be non-fatal.  One might want to try to connect to 500 PVs and accept some of them are not currently available and plow ahead.  It would be a shame if every `get()` needed to be within a try/except block -- that seems like Java, not Python....  I guess the question is really whether the user should have to be explicit about tolerating some unconnected PVs or can accept a value of `None` to mean "not available right now".   

For sure, we are not testing what happens with `ca.get()` on an unconnected channel (we were only testing what happens on `ca.connect()` for an unconnected channel).    So whatever the behavior was said to be, we were not doing a very good job of making sure it actually happened.  

I'm not excited about adding a flag to `ca.get()` about what to do in case of "trouble".  If the user is willing to set that flag,  couldn't they just use try/except? 

Anyway, a flag would not quite fix the problem here.  The problem here is that the code for `ca.put()` is using the withConnectedCHID decorator to ensure that it is passed a connected channel.  That means the code never actually got to `caget()` -- the error got trapped in the error-checking decorator. 

In short, I'm not opposed to changing the doc or the code for how to handle a `get()` of an unconnected PV.....  More thought and/or discussion needed!

**Comment by danielballan** [comment]
> This is our use case: we have a simple script that monitors the health of a number of IOCs

Could you issue an `EchoRequest` to each IOC and measure the time between the request and each server's `EchoResponse`? If I understand correctly, that is the protocol's way of supporting a health check, and if the IOC in question is overburdened at the moment, an echo imposes much less additional load than a batch of reads.

But as a more general question, it is interesting to considering how to express, "Do N things in parallel and collect the results as they complete, not allowing any particular failures to interrupt the rest." Python's built-in libraries illustrate one nice pattern that can be imitated. Python returns _futures_ upon which one can register callbacks, to use the results as they come in. As an example, here's how that pattern works out in caproto:

```python
from functools import partial
results = {}

def stash_result(name, response):
    results[name] = response.data

with Batch() as batch:
    for pv in pvs:
        batch.read(pv, partial(stash_result, pv.name))  # This queues up requests, but does not send them.
# Upon exit from `with` block, request(s) are sent efficiently in bulk, same as caget_many.
```

The `results` dict is populated with all the results, as they arrive. Any non-responses will simply never appear in the dict. (The above is lifted from [this section of the docs](https://nsls-ii.github.io/caproto/threading-client.html#batched-requests).)

Perhaps a similar functionality could be added to pyepics --- a new function `caget_many_tolerant` that blocks for up to `timeout` seconds and then returns whatever it has? By making this a separate function one avoids breaking public APIs.

**Comment by brunoseivam** [comment]
`caget_many`, if taken as described in the documentation, would be almost _exactly_ what we need. We reimplemented it in our script to work around the current issues. A check to `isConnected` was introduced to avoid a throwing `ca.get`. It is certainly not perfect (for example, `timeout` is used twice) but it works for the time being.

``` python
def caget_many(pvlist, as_string=False, count=None, as_numpy=True, timeout=1.0):
    chids, out = [], []
    for name in pvlist: 
        chids.append(ca.create_channel(name,auto_cb=False,connect=False))

    for chid in chids: 
        ca.connect_channel(chid)

    pvstatus=[ca.isConnected(chid) for chid in chids]
    t=time.time()
    while not all (pvstatus) and time.time()-t < timeout:
        ca.poll()
        pvstatus=[ca.isConnected(chid) for chid in chids]

    for chid,connected in zip(chids,pvstatus): 
        if connected:
            ca.get(chid, count=count, as_string=as_string, as_numpy=as_numpy, wait=False)

    for chid,connected in zip(chids,pvstatus): 
        if connected:
            out.append(ca.get_complete(chid,count=count,as_string=as_string,as_numpy=as_numpy,timeout=timeout))     
        else:
            out.append(None)

    return out
```

**Comment by newville** [comment]
@brunoseivam @danielballan   I think that Bruno's expected behavior of `caget_many()` is compelling.  It also seems like a common occurance to have not-currently-connected PVs.  So I'm inclined to follow the documentation and return `None` for an unconnected PV.    

It appears that this may have never actually worked as documented, but that surprises me.  I know that I played around with many variations of PV lists and remote networks for the code that is now in https://github.com/pyepics/pyepics/blob/master/tests/ca_fastconn.py and that led to `caget_many()`. So I would have sworn that it handled unconnected PVs well, even if that slowed it way down.   And, to be clear, this code does not use try/except to test for connected status.....

I don't see an advantage to raising an exception because a channel does not connect.  If the user wants to make a distinction between "it is OK if some PVs are unconnected"  and "it is NOT OK if some PVs are unconnected",  a value of `None` is a as clear as raising an exception.

    values = caget_many(pvlist)
    all_connected = None not in values

I'm not sure how a Batch() would help that, except that it handles the low-level I/O outside libca.   If I understand, an unconnected PV would give no entry in `results`, so the user would have to check that each PV in `pvs` had an entry in `results`, perhaps doing `value = results.get(pv, None)` or preloading values to be `None` with `results = {pv:None for pv in pvs}`.  I'm sure there are other approaches , but using `None` to mean "did not connect, no value known" seems OK to me too.

**Comment by danielballan** [comment]
I agree that including `None` in the results of a many-read provides the same information as leaving a key missing from a dictionariy, and in it that sense as clear. But when the caller takes those results and pushes them into some downstream function -- to format as text, update a GUI -- there may be an advantage to having only the "good" results. One avoids errors like "Cannot format None as float" if the calling code forgets to check `all_connected`, as calling code can often do. And if that consumer function expects a _specific_ PV to be in the results, it will raise a clear `KeyError` when it fails to find it, which the consumer can either catch (OK I don't _really_ need that) or propagate out ("Your Area Detector's Acquire PV is disconnected -- seems criticial!")

**Comment by danielballan** [comment]
I'll add that given that @brunoseivam and @newville seem to have consensus I'm perfectly happy to step back from the argument here and let this PR close to a resolution at this point -- just sharing my point of view.

**Comment by newville** [comment]
I think this is now documented better.
